File: get_ssh_key.py
#!/usr/bin/python

"""
get_ssh_key.py
Copyright 2010 Martin Carpenter, mcarpenter@free.fr.
Retrieves the current public SSH key from one or more remote hosts.
"""

import os.path
import paramiko
import socket
import sys
from optparse import OptionParser

DEFAULT_SSH_PORT = 22

def main(argv=None):
    """Main entry point."""
    if argv is None:
        argv = sys.argv
    ( key_types, hosts, port ) = parse_options(argv[1:])
    for host in hosts:
        key_str = get_host_key_str(key_types, host, port)
        if len(hosts) > 1:
            print "%s: %s" % ( host, key_str )
        else:
            print "%s" % ( key_str )

def parse_options(argv):
    """Parse the command line options. Exits on error with return
    code two, otherwise returns a list of key types, a list of
    host names and a port number."""
    parser = OptionParser("Usage: %prog [options] host [...]")
    parser.add_option('-t', '--type',
            help='specify key type TYPE (rsa or dsa)',
            metavar='TYPE')
    parser.add_option('-p', '--port',
            help='specify port number PORT',
            metavar='PORT')
    (options, hosts) = parser.parse_args(argv)
    if not options.port:
        port = DEFAULT_SSH_PORT
    else:
        if not options.port.isdigit():
            print >> sys.stderr, 'Invalid port number %s' % options.port
            parser.print_help()
            sys.exit(2)
        port = int( options.port )
    if not options.type:
        key_types = [ 'ssh-rsa', 'ssh-dss' ]
    elif options.type == 'rsa':
        key_types = [ 'ssh-rsa' ]
    elif options.type == 'dsa':
        key_types = [ 'ssh-dss' ]
    else:
        print >> sys.stderr, 'Invalid key type %s' % options.type
        parser.print_help()
        sys.exit(2)
    if not hosts:
        parser.print_help()
        sys.exit(2)
    return ( key_types, hosts, port )

def get_host_key_str(key_types, host, port):
    """Returns an ASCII string representing the host key."""
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((host, port))
    transport = paramiko.Transport(sock)
    options = transport.get_security_options()
    options.key_types = key_types
    transport.connect()
    key = transport.get_remote_server_key()
    transport.close()
    sock.close()
    return ( "%s %s %s#%s" %
            ( key.get_name(), key.get_base64(), host, key.get_bits()) )

if __name__ == "__main__":
    sys.exit(main())

File: get_ssh_key.rb
#!/usr/bin/ruby

# = get_ssh_key.rb
# Copyright 2010 Martin Carpenter, mcarpenter@free.fr.
#
# Retrieves the public SSH key from one or more remote hosts.
#
# == Usage
# get_ssh_key.rb -h | [ -t { rsa | dsa } ] host [...]

require 'rubygems'
require 'net/ssh'
require 'base64'
require 'getoptlong'

DEFAULT_SSH_PORT = 22

# Class to represent the server's public key. This must contain a
# #verify instance method since this callback is how we populate
# the @key attribute.
class Key

  # Return the length of the key in bits.
  def length
    case @key
    when OpenSSL::PKey::DSA
      length_indicator = @key.p
    when OpenSSL::PKey::RSA
      length_indicator = @key.n
    else
      raise RuntimeError, "Unknown key type `#{@key.class}'"
    end
    length_indicator.to_i.to_s(2).length
  end

  # Convert key to string representation.
  def to_s
    b64 = Base64.encode64(@key.to_blob).gsub!(/\s+/, '')
    "#{@key.ssh_type} #{b64}"
  end

  # Callback invoked by the host verifier.
  def verify(to_verify)
    @key = to_verify[:key]
  end

end

# Return the name of this program.
def program_name
  File.basename($0)
end

# Display usage on stderr and exit with (optional) exit code.
def usage(exit_code=0)
  $stderr.puts "Usage: #{program_name} -h | [ -t { rsa | dss } ] host [...]"
  exit(exit_code)
end

# Parse the command line options.
def parse_options
    opts = GetoptLong.new(
      [ '--help', '-h', GetoptLong::NO_ARGUMENT ],
      [ '--type', '-t', GetoptLong::REQUIRED_ARGUMENT ]
    )
  key_types = [ 'ssh-rsa', 'ssh-dss' ]
  begin
    opts.each do |opt, arg|
      case opt
      when '--help'
        usage
      when '--type'
        case arg
        when 'rsa'
          key_types = [ 'ssh-rsa' ]
        when 'dsa'
          key_types = [ 'ssh-dss' ]
        else
          $stderr.puts "Unknown key type `#{arg}'"
          usage(2)
        end
      end
    end
  rescue GetoptLong::MissingArgument, GetoptLong::InvalidOption
    usage(2)
  end
  usage(2) if ARGV.empty?
  [ key_types, ARGV ]
end

# Returns an ASCII string representing the host key.
def get_host_key_str(key_types, host)
  key = Key.new
  session = Net::SSH::Transport::Session.new(
    host,
    :port => DEFAULT_SSH_PORT,
    :paranoid => key,
    :host_key => key_types
  )
  session.close
  "#{key} #{host}##{key.length}"
end

key_types, hosts = *parse_options
hosts.each do |host|
  key_str = get_host_key_str(key_types, host)
  puts ( hosts.length > 1 ? "#{host}: #{key_str}" : key_str )
end

